File: src/test/java/com/github/loadtest4j/drivers/gatling/GatlingTest.java
package com.github.loadtest4j.drivers.gatling;

import com.github.loadtest4j.loadtest4j.driver.Driver;
import static com.github.loadtest4j.drivers.gatling.DriverResultAssert.assertThat;
import com.github.loadtest4j.loadtest4j.LoadTesterException;
import com.github.loadtest4j.loadtest4j.driver.DriverRequest;
import com.github.loadtest4j.loadtest4j.driver.DriverResult;
import com.xebialabs.restito.server.StubServer;
import org.glassfish.grizzly.http.Method;
import org.glassfish.grizzly.http.util.HttpStatus;
import org.junit.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

import static com.xebialabs.restito.builder.stub.StubHttp.whenHttp;
import static com.xebialabs.restito.builder.verify.VerifyHttp.verifyHttp;
import static com.xebialabs.restito.semantics.Action.status;
import static com.xebialabs.restito.semantics.Condition.*;
import org.junit.rules.ExpectedException;
import scala.concurrent.duration.FiniteDuration;

public class GatlingTest {
    private StubServer httpServer = new StubServer();

    @Before
    public void startServer() {
        httpServer.run();
    }

    @After
    public void stopServer() {
        httpServer.stop();
    }

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    private Driver sut() {
        final String serviceUrl = String.format("http://localhost:%d", httpServer.getPort());
        return new Gatling(new FiniteDuration(3, TimeUnit.SECONDS), serviceUrl, 1);
    }

    @Test
    public void testRun() {
        // Given
        final Driver driver = sut();
        // And
        whenHttp(httpServer).match(get("/")).then(status(HttpStatus.OK_200));

        // When
        final DriverResult result = driver.run(Collections.singletonList(DriverRequests.get("/")));

        // Then
        assertThat(result)
                .hasKo(0)
                .hasOkGreaterThan(1)
                .hasActualDurationGreaterThan(java.time.Duration.ZERO)
                .hasMaxResponseTimeGreaterThan(java.time.Duration.ZERO)
                .hasReportUrlWithScheme("file");
        // And
        verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/"));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testRunWithInvalidPercentile() {
        final Driver driver = sut();
        whenHttp(httpServer).match(get("/")).then(status(HttpStatus.OK_200));

        final DriverResult result = driver.run(Collections.singletonList(DriverRequests.get("/")));

        result.getResponseTime().getPercentile(102);
    }

    @Test
    public void testRunWithQueryString() {
        // Given
        final Driver driver = sut();
        // And
        whenHttp(httpServer).match(get("/")).then(status(HttpStatus.NOT_FOUND_404));
        // And
        whenHttp(httpServer).match(get("/"), parameter("foo", "bar")).then(status(HttpStatus.OK_200));

        // When
        final DriverResult result = driver.run(Collections.singletonList(DriverRequests.getWithQueryParams("/", Collections.singletonMap("foo", "bar"))));

        // Then
        assertThat(result)
                .hasOkGreaterThan(1)
                .hasKo(0);
    }

    @Test
    public void testRunWithElaborateRequest() {
        // Given
        final Driver driver = sut();
        // And
        whenHttp(httpServer)
                .match(post("/"), withHeader("Accept", "application/json"), withHeader("Content-Type", "application/json"), withPostBodyContaining("{}"))
                .then(status(HttpStatus.OK_200));

        // When
        final Map<String, String> headers = new ConcurrentHashMap<String, String>() {{
            put("Accept", "application/json");
            put("Content-Type", "application/json");
        }};
        final List<DriverRequest> requests = Collections.singletonList(DriverRequests.post("/", "{}", headers));
        final DriverResult result = driver.run(requests);

        // Then
        assertThat(result)
                .hasOkGreaterThan(1)
                .hasKo(0);
        // And
        verifyHttp(httpServer).atLeast(1, method(Method.POST), uri("/"), withHeader("Accept", "application/json"), withHeader("Content-Type", "application/json"), withPostBodyContaining("{}"));
    }

    @Test
    public void testRunWithMultipleRequests() {
        // Given
        final Driver driver = sut();
        // And
        whenHttp(httpServer).match(get("/")).then(status(HttpStatus.OK_200));
        // And
        whenHttp(httpServer).match(get("/pets")).then(status(HttpStatus.OK_200));

        // When
        final List<DriverRequest> requests = Arrays.asList(DriverRequests.get("/"), DriverRequests.get("/pets"));
        final DriverResult result = driver.run(requests);

        // Then
        assertThat(result)
                .hasOkGreaterThan(1)
                .hasKo(0);
        // And
        verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/"));
        // And
        verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/pets"));
    }

    @Test
    public void testRunWithNoRequests() {
        // Given
        final Driver driver = sut();

        // Expect
        thrown.expect(LoadTesterException.class);
        thrown.expectMessage("No requests were specified for the load test.");

        // When
        driver.run(Collections.emptyList());
    }

    static {
        // Silence Restito logging.
        Logger.getLogger("org.glassfish.grizzly").setLevel(Level.OFF);
    }
}

File: src/test/scala/com/github/loadtest4j/drivers/gatling/GatlingTest.scala
package com.github.loadtest4j.drivers.gatling

import java.util

import com.github.loadtest4j.loadtest4j.{Driver, LoadTesterException}
import com.xebialabs.restito.server.StubServer
import org.glassfish.grizzly.http.Method
import org.glassfish.grizzly.http.util.HttpStatus
import org.junit._
import java.util.Collections
import java.util.logging._

import com.xebialabs.restito.builder.stub.StubHttp.whenHttp
import com.xebialabs.restito.builder.verify.VerifyHttp.verifyHttp
import com.xebialabs.restito.semantics.Action.status
import com.xebialabs.restito.semantics.Condition._
import org.junit.Assert._
import scala.util.{Try, Failure}
import scala.concurrent.duration._

class GatlingTest {

  private val httpServer = new StubServer()

  @Before
  def startServer(): Unit = {
    httpServer.run()
  }

  @After
  def stopServer(): Unit = {
    httpServer.stop()
  }

  private def sut(): Driver = {
    val serviceUrl = "http://localhost:%d".format(httpServer.getPort)
    new Gatling(3.seconds, serviceUrl, 1)
  }

  @Test
  def testRun(): Unit = {
    // Given
    val driver = sut()
    // And
    whenHttp(httpServer).`match`(get("/")).`then`(status(HttpStatus.OK_200))

    // When
    val result = driver.run(Collections.singletonList(DriverRequests.get("/")))

    // Then
    assertEquals(0, result.getKo)
    assertGreaterThanOrEqualTo(1, result.getOk)
    assertTrue(result.getActualDuration.toMillis > 0)
    assertStartsWith("file://", result.getReportUrl.get())
    assertEndsWith("html", result.getReportUrl.get())
    // And
    verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/"))
  }

  @Test
  def testRunWithElaborateRequest(): Unit = {
    // Given
    val driver = sut()
    // And
    whenHttp(httpServer)
      .`match`(post("/"), withHeader("Accept", "application/json"), withHeader("Content-Type", "application/json"), withPostBodyContaining("{}"))
      .`then`(status(HttpStatus.OK_200))

    // When
    val requests = Collections.singletonList(DriverRequests.post("/", "{}", Map("Accept" -> "application/json", "Content-Type" -> "application/json")))
    val result = driver.run(requests)

    // Then
    assertEquals(0, result.getKo)
    assertGreaterThanOrEqualTo(1, result.getOk)
    // And
    verifyHttp(httpServer).atLeast(1, method(Method.POST), uri("/"), withHeader("Accept", "application/json"), withHeader("Content-Type", "application/json"), withPostBodyContaining("{}"))
  }

  @Test
  def testRunWithMultipleRequests(): Unit = {
    // Given
    val driver = sut()
    // And
    whenHttp(httpServer).`match`(get("/")).`then`(status(HttpStatus.OK_200))
    // And
    whenHttp(httpServer).`match`(get("/pets")).`then`(status(HttpStatus.OK_200))

    // When
    val requests = util.Arrays.asList(DriverRequests.get("/"), DriverRequests.get("/pets"))
    val result = driver.run(requests)

    // Then
    assertGreaterThanOrEqualTo(1, result.getOk)
    assertEquals(0, result.getKo)
    // And
    verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/"))
    // And
    verifyHttp(httpServer).atLeast(1, method(Method.GET), uri("/pets"))
  }

  @Test
  def testRunWithNoRequests(): Unit = {
    // Given
    val driver = sut()

    // When
    Try(driver.run(Collections.emptyList())) match {
      case Failure(e: LoadTesterException) => assertEquals("No requests were specified for the load test.", e.getMessage)
      case _ => fail("This should not work.")
    }
  }

  private def assertStartsWith(prefix: String, actual: String): Unit = {
    val msg = "'%s' did not start with the substring '%s'.".format(actual, prefix)
    assertTrue(msg, actual.startsWith(prefix))
  }

  private def assertEndsWith(suffix: String, actual: String): Unit = {
    val msg = "'%s' did not end with the substring '%s'.".format(actual, suffix)
    assertTrue(msg, actual.endsWith(suffix))
  }

  private def assertGreaterThanOrEqualTo(expected: Long, actual: Long): Unit  ={
    val msg = "Expected %d to be >= %d, but it was not.".format(actual, expected)
    assertTrue(msg, actual >= expected)
  }
}

object GatlingTest {
  // Silence Restito logging.
  Logger.getLogger("org.glassfish.grizzly").setLevel(Level.OFF)
}